crates/core/src/lib.rs

```diff
@@ -36,46 +186,68 @@ pub extern "C" fn init() {
         let main = suborbital.get_property("run_e").unwrap();
+        let env = suborbital.get_property("env").unwrap();
+
+        // Unlike the other language bindings, JS never considers itself a
+        // WebAssembly module (which makes sense—JS is usually interoping with
+        // wasm modules, not compiled to wasm itself). For this reason, the
+        // bindings look at the host as though it's a wasm module. Even though
+        // we're importing host functions, the JS bindings will consider them
+        // exports from a wasm module. This doesn't have a major effect other
+        // than some confusing terminology.
+        let imports = context.object_value().unwrap();
+        setup_imports(&context, &imports);
+
+        // The JS expects our host functions under the `_exports` key on `env`.
+        env.set_property("_exports", imports)
+            .expect("failed to set _exports on env");
 
         JS_CONTEXT.set(context).unwrap();
         ENTRYPOINT.0.set(suborbital).unwrap();
         ENTRYPOINT.1.set(main).unwrap();
     }
 }
 
-struct JsRunnable {}
-
-impl Runnable for JsRunnable {
-    fn run(&self, in_bytes: Vec<u8>) -> Result<Vec<u8>, RunErr> {
-        unsafe {
-            let context = JS_CONTEXT.get().unwrap();
-            let suborbital = ENTRYPOINT.0.get().unwrap();
-            let main = ENTRYPOINT.1.get().unwrap();
-            let input = context
-                .value_from_bytes(in_bytes)
-                .expect("Couldn't load input");
-            let output_value = main.call(&suborbital, &[input]);
-            if output_value.is_err() {
-                panic!("{}", output_value.unwrap_err().to_string());
-            }
-            match output_value {
-                Ok(value) => {
-                    let vec = context.value_to_bytes(value).unwrap();
-                    Ok(vec)
-                }
-                Err(err) => {
-                    let message = err.to_string();
-                    Err(RunErr { code: 500, message })
-                }
-            }
+/// # Safety
+#[no_mangle]
+pub unsafe extern "C" fn run_e(pointer: *mut u8, size: i32, ident: i32) {
+    extern "C" {
+        fn return_error(code: i32, result_pointer: *const u8, result_size: i32, ident: i32);
+    }
+
+    let in_bytes = Vec::from_raw_parts(pointer, size as usize, size as usize);
+
+    let context = JS_CONTEXT.get().unwrap();
+    let suborbital = ENTRYPOINT.0.get().unwrap();
+    let main = ENTRYPOINT.1.get().unwrap();
+    let input = context
+        .value_from_bytes(in_bytes)
+        .expect("Couldn't load input");
+    let id = context.value_from_i32(ident).expect("Couldn't load ident");
+    let output_value = main.call(&suborbital, &[input, id]);
+
+    match output_value {
+        Ok(_) => (),
+        Err(err) => {
+            let message = err.to_string();
+            return_error(500, message.as_ptr(), message.len() as i32, ident);
```

Comment:
Will we ever allow the JS code to dictate the error code?

Reply:
They can, just by calling `returnError` from the JS. This case is just for any uncaught JS errors that occur. Even if the module calls `returnError`, it'll hit the `Ok` case above since there was no uncaught JS error.